### PR TITLE
Use a circular buffer for logging during unittests

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Using Cycle.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
So now when there is a failure, you will see logs like this: https://travis-ci.com/bpfkorea/agora/jobs/247032014

Rebased to go green.

File: source/agora/utils/Log.d [hunk 1]
```diff
@@ -54,15 +54,94 @@ public template AddLogger (string moduleName = __MODULE__)
...
-        appender.layout(new AgoraLayout());
-        Log.root.add(appender);
+
+    appender.layout(new AgoraLayout());
+    Log.root.add(appender);
+}
+
+import ocean.transition;
+import ocean.util.log.Appender;
+import ocean.util.log.Event;
```
Comment (Geod24) [inline source/agora/utils/Log.d]: Move those with other imports

File: source/agora/utils/Log.d [hunk 2]
```diff
@@ -54,15 +54,94 @@ public template AddLogger (string moduleName = __MODULE__)
...
+        auto prev_idx = (this.buffer_idx + 1) % LineLimit;
+        while (prev_idx != this.buffer_idx)
+        {
+            if (this.line_buffer[prev_idx].length > 0)
+                writeln(this.line_buffer[prev_idx]);
+
+            prev_idx = (prev_idx + 1) % LineLimit;
+        }
+
+        auto last_idx = this.buffer_idx;
```
Comment (Geod24) [inline source/agora/utils/Log.d]: Dead code ?

**Review by Geod24 — APPROVED** [review]
Looks pretty neat

File: source/agora/test/Base.d [hunk 3]
```diff
@@ -215,6 +218,24 @@ public class TestAPIManager
...
+
+    ***************************************************************************/
+
+    public void printLogs ()
+    {
+        import std.stdio;
+        foreach (key, api; this.apis)
+        {
+            writefln("Log for node %s:", key);
+            writeln ("======================================================================");
```
Comment (Geod24) [inline source/agora/test/Base.d]: Style

**Comment by Geod24** [comment]
> So now when there is a failure, you will see logs like this: https://travis-ci.com/bpfkorea/agora/jobs/247032014

RAD

**Comment by AndrejMitrovic** [comment]
Fixed styling issues.

File: source/agora/utils/Log.d [hunk 4]
```diff
@@ -54,15 +55,88 @@ public template AddLogger (string moduleName = __MODULE__)
...
+    private Mask mask_;
+
+    ///
+    private const LineLimit = 100;
+
+    ///
+    private size_t buffer_idx;
+
+    ///
+    private mstring[LineLimit] line_buffer;
```
Comment (Geod24) [inline source/agora/utils/Log.d]: Note: Originally, when I meant circular buffer, I was thinking of just a buffer of 64k. The idea was that GC-allocation can slow things down and hide bugs (see [this](https://preshing.com/20120522/lightweight-in-memory-logging/)).
I don't think it's much of a concern for us at the moment, or ever will be, just wanted to add perspective.

**Review by Geod24 — APPROVED** [review]
Feel free to merge when green.

File: source/agora/utils/Log.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/agora/utils/Log.d]: Right. I was worried the output would be cut mid-sentence and might cause misleading error logs. But 64k should be big enough.

Allow me to rework the PR? :)

**Comment by AndrejMitrovic** [comment]
Updated. It's simplified and more efficient now.

**Review by Geod24 — APPROVED** [review]
Wow, TIL about `cyclic`. Output still the same ?

**Comment by AndrejMitrovic** [comment]
Pushed a fix for an edge-case with printing when the buffer is not exhausted yet. Added a comment too.

**Pull request merged**